stan/math/rev/arr/functor/coupled_ode_system.hpp

```diff
@@ -443,15 +442,16 @@ struct coupled_ode_system<F, var, var> {
                      const std::vector<int>& x_int, std::ostream* msgs)
       : f_(f),
         y0_(y0),
-        y0_dbl_(value_of(y0)),
         theta_(theta),
-        theta_dbl_(value_of(theta)),
         x_(x),
         x_int_(x_int),
         N_(y0.size()),
         M_(theta.size()),
         size_(N_ + N_ * (N_ + M_)),
-        msgs_(msgs) {}
+        msgs_(msgs) {
+    for (auto& p : theta)
+      theta_copy_.emplace_back(var(new vari(value_of(p), false)));
```

Comment:
If we know `p` is a `stan::math::var`, why not use `v.val()` directly? I think it'd be a little clearer.

Reply:
To me the `value_of` is actually clearer as I don't need to recal the details of the `var` class. Calling `value_of` always gives me a double - no matter what. The `val` method feels like an internal function whereas the `value_of` is a clear API.

Reply:
I'll try to explain why I find this so much more confusing than if this was:
```
for (var&& p : theta) 
  theta_copy_.emplace_back(var(new vari(p.val(), false));
```

And this is coming from me having to read this code to figure it out.
1. `value_of()` is a function that's designed to work on `double`, `var`, `fvar<double>`, `fvar<var>`, and `fvar<fvar<var>>`. So I'm looking at `value_of(p)` and trying to figure out... "what is p?"
2. `auto& p` instead of `var&& p` doesn't help me at all. I know that `p` is one of `theta`, so `theta` is a collection, but what is it? Since I'm flipping between base template classes that can handle all autodiff types, it's not always easy to keep track that this is a `rev` only implementation. It'd really just help to have this be `var&& p` so the type is obvious.
3. `theta`... now I need to go and find where this is defined. C++ is a structured language, but it actually allows the declaration of struct and class member variables either at the top, below, or anywhere in between. And now I've got to search. I was looking at this using the GitHub interface, so the actual declaration was off of the interface; in order for me to go look at this, I had to go and check out the branch, open the file, all to find that `theta` is declared as `std::vector<var>`.

I think the thing that we're not seeing eye-to-eye on is the use of `value_of`. I find it much easier to reason about something that's only valid for `var` instead of having to work to figure out whether the object passed in is one of many different types. (It's just a little more mental work the reader has to do when it can all be made explicit as an alternative.)

(By the way, your statement is incorrect -- `value_of` does not always give a `double`. It'll give the type of the `value`, which for `fvar<var>` will be `var`.)

Hopefully you've gotten some context on why this is harder to read for a reviewer or anyone else coming to the code later on. This is minor, but the idea still stands -- make it easier for the next person to read. It takes effort now, but it's so worth it later on.

Reply:
Oh... `value_of` is defined differently in my head... I always forget about higher order things. With the correct definitions the `val` call is better, yeah. Sorry, but I am not used to meaning of `value_of`.

Honestly, I do expect the reviewer to look at the file and not just the diff from git. It is really hard to expect our changes to be self-explanatory on a single diff line. The auto vs actual type ... I read the opposite elsewhere. Your third point I do not quite understand. I am looping over theta which is the argument to the constructor.

Reply:
Thanks! The third point wasn't actually there... you're right. I should have looked at the constructor signature, not the actual type of the variable.